skbio/util/_decorator.py

```diff
@@ -82,3 +86,252 @@ def __get__(self, cls, owner):
 
     def __set__(self, obj, value):
         raise AttributeError("can't set attribute")
+
+
+class _state_decorator(object):
+    """ Base class for decorators of all public functionality.
+    """
+
+    _required_kwargs = ()
+
+    def _get_indentation_level(self, docstring, default_existing_docstring=4,
+                               default_no_existing_docstring=0):
+        """ Determine the level of indentation of the docstring to match it.
+
+            The indented content after the first line of a docstring can
+            differ based on the nesting of the functionality being documented.
+            For example, a top-level function may have its "Parameters" section
+            indented four-spaces, but a method nested under a class may have
+            its "Parameters" section indented eight spaces. This function
+            determines the indentation level of the first non-whitespace line
+            following the initial summary line.
+        """
+        # if there is no existing docstring, return the corresponding default
+        if docstring is None:
+            return default_no_existing_docstring
+
+        docstring_lines = docstring.split('\n')
+
+        # if there is an existing docstring with only a single line, return
+        # the corresponding default
+        if len(docstring_lines) == 1:
+            return default_existing_docstring
+
+        # find the first non-blank line (after the initial summary line) and
+        # return the number of leading spaces on that line
+        for line in docstring_lines[1:]:
+            if len(line.strip()) == 0:
+                # ignore blank lines
+                continue
+            else:
+                return len(line) - len(line.lstrip())
+
+        # if there is an existing docstring with only a single non-whitespace
+        # line, return the corresponding default
+        return default_existing_docstring
+
+    def _update_docstring(self, docstring, state_desc,
+                          state_desc_prefix='State: '):
+        docstring_content_indentation = \
+            self._get_indentation_level(docstring)
+        # Hande the case of no initial docstring
+        if docstring is None:
+            return "%s%s%s" % (
+                ' ' * docstring_content_indentation, state_desc_prefix,
+                state_desc)
+
+        doc_lines = docstring.split('\n')
```

Comment:
You're splitting the docstring twice, here and in `_get_indentation_level`. Can you refactor to avoid performing this operation twice? I'm mainly concerned about efficiency, not so much code duplication.